packages/sanity/src/core/releases/tool/components/releaseCTAButtons/ReleaseRevertButton/useDocumentRevertStates.ts

```diff
@@ -0,0 +1,135 @@
+import {type SanityDocument} from '@sanity/types'
+import {useCallback, useEffect, useMemo, useRef} from 'react'
+import {useObservable} from 'react-rx'
+import {catchError, filter, forkJoin, from, map, type Observable, of, switchMap} from 'rxjs'
+
+import {useClient} from '../../../../../hooks/useClient'
+import {getTransactionsLogs} from '../../../../../store/translog/getTransactionLogs'
+import {API_VERSION} from '../../../../../tasks/constants'
+import {type DocumentInRelease} from '../../../detail/useBundleDocuments'
+
+export type RevertDocument = SanityDocument & {
+  _system?: {
+    delete: boolean
+  }
+}
+
+type RevertDocuments = RevertDocument[]
+
+type DocumentRevertStates = RevertDocuments | null | undefined
+
+export const useDocumentRevertStates = (documents: DocumentInRelease[]) => {
+  const client = useClient({apiVersion: API_VERSION})
+  const observableClient = client.observable
+  const transactionId = documents[0]?.document._rev
+  const {dataset} = client.config()
+
+  const resultPromiseRef = useRef<Promise<DocumentRevertStates> | null>(null)
+  const resolvedDocumentRevertStatesPromiseRef = useRef<
+    ((value: DocumentRevertStates) => void) | null
+  >(null)
+  const resolvedDocumentRevertStatesResultRef = useRef<DocumentRevertStates | null>(null)
+
+  useEffect(() => {
+    if (!resultPromiseRef.current) {
+      resultPromiseRef.current = new Promise((resolve) => {
+        resolvedDocumentRevertStatesPromiseRef.current = resolve
```

Comment:
This looks like a good case for [Promise.withResolvers()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers)